### PR TITLE
Add Caching for Issuer Data

### PR DESCRIPTION
**Pull request by Sneagan**

### CODE REVIEW

**Review by husobee — DISMISSED** [review]

**Review by johnhalbert — APPROVED** [review]

**Comment by Sneagan** [comment]
This expanded into a larger project when I realized that what looked like a caching implementation actually didn't work. This PR should now include caching for issuer data as fetched from the `server` package.

File: server/issuers.go [hunk 1]
```diff
@@ -83,6 +83,7 @@ func (c *Server) GetLatestIssuerKafka(issuerType string, issuerCohort int16) (*I
 }
 
 // GetIssuers - get all issuers by issuer type
+// @TODO: This function is no longer needed. Delete.
```
Comment (husobee) [inline server/issuers.go:86]: Is  this comment true for the V1 Issuers in grants?

**Review by husobee — DISMISSED** [review]
Looks good, but I hesitate to add the delete comments as todos as they may be used by grants  v1 issuer use cases.

File: server/db.go [hunk 2]
```diff
@@ -448,17 +442,15 @@ func (c *Server) fetchIssuerByType(ctx context.Context, issuerType string) (*Iss
...
 	return convertedIssuer, nil
 }
 
 func (c *Server) fetchIssuers(issuerType string) (*[]Issuer, error) {
-	if c.caches != nil {
-		if cached, found := c.caches["issuers"].Get(issuerType); found {
-			return cached.(*[]Issuer), nil
-		}
+	if cached := retrieveFromCache(c.caches, "issuers", issuerType); cached != nil {
+		return cached.(*[]Issuer), nil
```
Comment (johnhalbert) [inline server/db.go:453]: Do we need to check here that the type assertion will pass and we have one or more issuers in the slice?

File: server/db.go [hunk 3]
```diff
@@ -1220,3 +1190,26 @@ func isPostgresNotFoundError(err error) bool {
...
+) interface{} {
+	if caches != nil {
+		if cached, found := caches[cacheName].Get(key); found {
+			return cached
+		}
+	}
+	return nil
+}
+
+func bootstraCache(cfg DBConfig) map[string]CacheInterface {
```
Comment (johnhalbert) [inline server/db.go:1207]: nit: `bootstrapCache`?

**Review by johnhalbert — APPROVED** [review]
lgmt. Raised a question about how we handle cached values for `fetchIssuers`, but it doesn't appear to be called anywhere, so shouldn't block PR.

File: server/db.go (hunk 2 above)
Comment (IanKrieger) [reply server/db.go:453]: To this point, slices are already allowed to be nil w/o needing a pointer. Maybe this is for a different day, but if it's easy enough can you have it not be a pointer to a slice?

File: server/issuers.go (hunk 1 above)
Comment (Sneagan) [reply server/issuers.go:86]: Fair enough. I'll change to recommend refactor or deprecation, as the functions are inefficient even we still need them now that there are not multiple issuers per cohort.

File: server/db.go (hunk 2 above)
Comment (Sneagan) [reply server/db.go:453]: I thought I got all of these. Good catch. Fixing.

**Review by johnhalbert — APPROVED** [review]

File: server/db.go [hunk 4]
```diff
@@ -448,16 +442,16 @@ func (c *Server) fetchIssuerByType(ctx context.Context, issuerType string) (*Iss
...
 }
 
-func (c *Server) fetchIssuers(issuerType string) (*[]Issuer, error) {
-	if c.caches != nil {
-		if cached, found := c.caches["issuers"].Get(issuerType); found {
-			return cached.(*[]Issuer), nil
+func (c *Server) fetchIssuers(issuerType string) ([]Issuer, error) {
+	if cached := retrieveFromCache(c.caches, "issuers", issuerType); cached != nil {
+		if issuers, ok := cached.([]Issuer); ok {
+			return issuers, nil
```
Comment (johnhalbert) [inline server/db.go:454]: One other thing here:

```
var nilslc []string
fmt.Println([]string{} == nil) // false
fmt.Println(nilslc == nil) //true
```

I have a feeling we're not going to cache an empty slice, but it's not the same as a nil slice, so the check `cached != nil` could allow the return of an empty slice.

**Review by johnhalbert — APPROVED** [review]

File: server/db.go (hunk 4 above)
Comment (Sneagan) [reply server/db.go:454]: This function really will only ever return a slice of 1 issuer. I think we can allow this through and come back to refactor the slice issue generally.

**Pull request merged**